### PR TITLE
AIRFLOW-2068 MesosExecutor allows specifying optional Docker

### PR DESCRIPTION
**Pull request by agrajm**
AIRFLOW-2068 Mesos Executor should allow specifying optional Docker image before running airflow command

In its current form, MesosExecutor schedules tasks on mesos slaves which just contain airflow commands assuming that the mesos slaves already have airflow installed and configured on them. This assumption goes against the Mesos philosophy of having a heterogeneous cluster.

Since Mesos provides an option to pull a docker image before actually running the actual task/command so this improvement changes the mesos_executor.py to specify an optional docker image containing airflow which can be pulled on slaves before running the actual airflow command.

### CODE REVIEW

**Comment by agrajm** [comment]
Please take a look @mistercrunch @kapil-malik

**Comment by bolkedebruin** [comment]
Great addition. Can you please add tests to cover this?

**Comment by agrajm** [comment]
sure will add tests and update the PR

**Comment by agrajm** [comment]
@bolkedebruin I've added a test for mesos executor. Please have a look.

**Comment by bolkedebruin** [comment]
The test is skipped at the moment. Can you make sure it runs when on Travis?

**Comment by agrajm** [comment]
@bolkedebruin The test is skipped as it requires Mesos python bindings/eggs to run. I tested it locally and it works fine. See below:

```
(env) ➜  incubator-airflow git:(AIRFLOW-2068) ✗ ./run_unit_tests.sh tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor -s --logging-level=DEBUG
+ export AIRFLOW_HOME=/usr/local/airflow
+ AIRFLOW_HOME=/usr/local/airflow
+ export AIRFLOW__CORE__UNIT_TEST_MODE=True
+ AIRFLOW__CORE__UNIT_TEST_MODE=True
+ export AIRFLOW__TESTSECTION__TESTKEY=testvalue
+ AIRFLOW__TESTSECTION__TESTKEY=testvalue
+ export AIRFLOW_USE_NEW_IMPORTS=1
+ AIRFLOW_USE_NEW_IMPORTS=1
+++ dirname ./run_unit_tests.sh
++ cd .
++ pwd
+ DIR=/Users/amangal/github/agrajm/incubator-airflow
+ export PYTHONPATH=:/Users/amangal/github/agrajm/incubator-airflow:/Users/amangal/github/agrajm/incubator-airflow/tests/test_utils
+ PYTHONPATH=:/Users/amangal/github/agrajm/incubator-airflow:/Users/amangal/github/agrajm/incubator-airflow/tests/test_utils
+ nose_args='tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor -s --logging-level=DEBUG'
+ which airflow
+ echo 'Initializing the DB'
Initializing the DB
+ yes
+ airflow resetdb
[2018-02-09 21:48:58,090] {__init__.py:45} INFO - Using executor SequentialExecutor
DB: sqlite:////usr/local/airflow/unittests.db
This will drop existing tables if they exist. Proceed? (y/n)[2018-02-09 21:48:58,270] {db.py:319} INFO - Dropping tables that exist
[2018-02-09 21:48:58,299] {migration.py:116} INFO - Context impl SQLiteImpl.
[2018-02-09 21:48:58,299] {migration.py:121} INFO - Will assume non-transactional DDL.
[2018-02-09 21:48:58,306] {db.py:300} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> e3a246e0dc1, current schema
INFO  [alembic.runtime.migration] Running upgrade e3a246e0dc1 -> 1507a7289a2f, create is_encrypted
/Users/amangal/github/agrajm/incubator-airflow/env/lib/python2.7/site-packages/alembic/util/messaging.py:69: UserWarning: Skipping unsupported ALTER for creation of implicit constraint
  warnings.warn(msg)
INFO  [alembic.runtime.migration] Running upgrade 1507a7289a2f -> 13eb55f81627, maintain history for compatibility with earlier migrations
INFO  [alembic.runtime.migration] Running upgrade 13eb55f81627 -> 338e90f54d61, More logging into task_isntance
INFO  [alembic.runtime.migration] Running upgrade 338e90f54d61 -> 52d714495f0, job_id indices
INFO  [alembic.runtime.migration] Running upgrade 52d714495f0 -> 502898887f84, Adding extra to Log
INFO  [alembic.runtime.migration] Running upgrade 502898887f84 -> 1b38cef5b76e, add dagrun
INFO  [alembic.runtime.migration] Running upgrade 1b38cef5b76e -> 2e541a1dcfed, task_duration
INFO  [alembic.runtime.migration] Running upgrade 2e541a1dcfed -> 40e67319e3a9, dagrun_config
INFO  [alembic.runtime.migration] Running upgrade 40e67319e3a9 -> 561833c1c74b, add password column to user
INFO  [alembic.runtime.migration] Running upgrade 561833c1c74b -> 4446e08588, dagrun start end
INFO  [alembic.runtime.migration] Running upgrade 4446e08588 -> bbc73705a13e, Add notification_sent column to sla_miss
INFO  [alembic.runtime.migration] Running upgrade bbc73705a13e -> bba5a7cfc896, Add a column to track the encryption state of the 'Extra' field in connection
INFO  [alembic.runtime.migration] Running upgrade bba5a7cfc896 -> 1968acfc09e3, add is_encrypted column to variable table
INFO  [alembic.runtime.migration] Running upgrade 1968acfc09e3 -> 2e82aab8ef20, rename user table
INFO  [alembic.runtime.migration] Running upgrade 2e82aab8ef20 -> 211e584da130, add TI state index
INFO  [alembic.runtime.migration] Running upgrade 211e584da130 -> 64de9cddf6c9, add task fails journal table
INFO  [alembic.runtime.migration] Running upgrade 64de9cddf6c9 -> f2ca10b85618, add dag_stats table
INFO  [alembic.runtime.migration] Running upgrade f2ca10b85618 -> 4addfa1236f1, Add fractional seconds to mysql tables
INFO  [alembic.runtime.migration] Running upgrade 4addfa1236f1 -> 8504051e801b, xcom dag task indices
INFO  [alembic.runtime.migration] Running upgrade 8504051e801b -> 5e7d17757c7a, add pid field to TaskInstance
INFO  [alembic.runtime.migration] Running upgrade 5e7d17757c7a -> 127d2bf2dfa7, Add dag_id/state index on dag_run table
INFO  [alembic.runtime.migration] Running upgrade 127d2bf2dfa7 -> cc1e65623dc7, add max tries column to task instance
INFO  [alembic.runtime.migration] Running upgrade cc1e65623dc7 -> bdaa763e6c56, Make xcom value column a large binary
INFO  [alembic.runtime.migration] Running upgrade bdaa763e6c56 -> 947454bf1dff, add ti job_id index
INFO  [alembic.runtime.migration] Running upgrade 947454bf1dff -> d2ae31099d61, Increase text size for MySQL (not relevant for other DBs' text types)
INFO  [alembic.runtime.migration] Running upgrade d2ae31099d61 -> 0e2a74e0fc9f, Add time zone awareness
+ airflow initdb
[2018-02-09 21:49:00,329] {__init__.py:45} INFO - Using executor SequentialExecutor
DB: sqlite:////usr/local/airflow/unittests.db
[2018-02-09 21:49:00,446] {db.py:300} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
Done.
+ '[' '' ']'
+ '[' -z 'tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor -s --logging-level=DEBUG' ']'
+ echo 'Starting the unit tests with the following nose arguments: tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor' -s --logging-level=DEBUG
Starting the unit tests with the following nose arguments: tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor -s --logging-level=DEBUG
+ nosetests tests.contrib.executors.test_mesos_executor:MesosExecutorTest.test_mesos_executor -s --logging-level=DEBUG
[2018-02-09 21:49:01,915] {__init__.py:45} INFO - Using executor SequentialExecutor
[2018-02-09 21:49:02,902] {mesos_executor.py:124} INFO - Received offer 1 with cpus: 2 and mem: 4
[2018-02-09 21:49:02,902] {mesos_executor.py:137} INFO - Launching task 0 using offer 1
.
----------------------------------------------------------------------
Ran 1 test in 0.005s

OK
[2018-02-09 21:49:02,903] {settings.py:165} DEBUG - Disposing DB connection pool (PID 47090)
(env) ➜  incubator-airflow git:(AIRFLOW-2068) 
```

>>Can you make sure it runs when on Travis?

Not sure how to do this. I did run the test successfully on my setup though

**Comment by agrajm** [comment]
Hi @bolkedebruin could you please merge this one if everything looks on track?

**Comment by bolkedebruin** [comment]
Thanks, but we can extend travis to include mesos. This will help you in the future as well as we can ensure any follow up changes do not break. So can you please add the mesos dependencies and make sure the tests are activated on Travis?

**Comment by agrajm** [comment]
@bolkedebruin I totally agree with your thoughts but the problem with extending travis to include mesos is that I would have to build mesos from source since the mesos python bindings are not `pip` installable, neither the python eggs are available for direct download which can be then `easy_install`ed. I've opened an issue in Mesos for making those eggs available for downlaod (https://issues.apache.org/jira/browse/MESOS-8597) to resolve that. 

Do you want me to build mesos from source and then extract eggs out of it? It takes approximately half an hour on my macbook Pro. Not sure about the configuration of travis build machine but If we can afford to add that much time to the travis build, please let me know.

**Comment by sebradloff** [comment]
Very excited to see this PR closed and shipped. It's something that I've been struggling with for the past week to get the MesosExecutor setup without going against the Mesos philosophy. @agrajm if you need any help with documentation on it's use, I'd be more than happy to help.

**Comment by agrajm** [comment]
@sebradloff funny you should mention it, I was writing a small blog post about how to enable this & just finished wrapping it up - http://agrajmangal.in/blog/big-data/mesos-airflow-docker/ Feel free to suggest changes and/or improvements

**Comment by sebradloff** [comment]
@agrajm Love the post! 

To deal with the Mesos python eggs issue, we've built an upstream docker image for the version of Mesos we use (1.4.0) and then simply use the docker multi stage build to copy over those eggs from that image. We plan on getting that together and releasing the pattern within the next 2 weeks once we've got the MesosExecutor running.

Regarding the `docker_image_slave` setting, it would be really cool if we could also specify memory and CPU constraints on a per Operator or DAG basis so that we're optimizing our resources to the fullest. Correct me if I'm wrong, this is my interpretation based on the source code. Full disclosure we haven't run the MesosExecutor yet but just confirmed running the LocalExecutor in a deployed environment with s3 remote logs and external psql metadabase.

Totally understand the need to specify docker as a containerizer for Mesos agents, but what's the reasoning behind needing `mesos` listed as a conatinerizer? Reading [the docs](http://mesos.apache.org/documentation/latest/containerizers/)  I'm not sure I follow how specifically the Airflow MesosExecutor setup would require the `mesos` containerizer assuming Docker provides all the control you would need.

I think an amazing addition to that post would be an architecture diagram of what all this looks like. I took a shot at it here: https://drive.google.com/file/d/1sfbZ7W14BeYvqoOCupvB8vADuDQY1Xmt/view Basically something that gives a picture of the architecture that you need to standup and making it clear that tasks are spun up dynamically in containers.

Please lmk if there's anything else I can help with. Really want this to succeed and help make this all easier to understand for those that come along. Overall, I can feel the shift of infrastructure people being tasked with operationalizing airflow in more than just "one massive EC2 instance". Anything we can do to make this workflow more easily understood will be a win for data scientists and infrastructure teams.

**Comment by agrajm** [comment]
@sebradloff Sorry for late reply. 

In its current form, the MesosExecutor allows specifying static values for CPU & memory per task. When the airflow framework scheduler receives a resource offer from mesos master that has less resources than what is specified in this config, it rejects that offer. Or in other words, it accepts only those offers that have more or equal number of resources as mentioned in the config. But its a STATIC value. Same for all tasks for all DAGs. 
Now I'll preface this by saying I'm pretty new to airflow and specifically to the scheduler-foo logic but FWIK having able to specify cpu & memory per task for a DAG would require changes in airflow scheduler's code where it's creating `TaskInstance`s but it sounds like a suggestion worth exploring. 

Regarding having to specify both `docker` & `mesos` containerizers it looks like you are right. We need only one in this case - the `docker` one. Although I'm yet to try that out on my setup. Not getting enough bandwidth :( 

Thanks for the architectural diagram. I did create a [basic architectural diagram](http://agrajmangal.in/blog/wp-content/uploads/2018/01/airflow_mesos.png) in [my first blog post](http://agrajmangal.in/blog/big-data/running-airflow-on-top-of-apache-mesos/) on this subject - See http://agrajmangal.in/blog/wp-content/uploads/2018/01/airflow_mesos.png But I agree specifying more details like containers inside executors and webserver and scheduler also running on mesos slaves makes sense. Will update the blog post as soon as I find some time.

**Comment by sebradloff** [comment]
@agrajm No worries on the late reply.

Your first point regarding CPU and memory per task makes sense. I think I will also have to do a deeper dive into airflow to understand how we could potentially change the amount of resources. 

Thanks for confirming my suspicions on the need for both `docker` and `mesos` containerizers.

Yea, your first blog post was great and opened up our eyes to the possibility of Mesos and airflow to begin with. The diagram was the confusing part from the sense of what was a container vs a VM.

One last thing, just to confirm my understanding. A new offer will be made to Mesos for each airflow task, in the code that you've written here, correct? I'm basing that off of your test code where you create two fake tasks and I see them being added to a tasks queue with the scheduler creating offers plural.

**Comment by agrajm** [comment]
So let's see this from mesos POV - Airflow is just one framework registered with mesos amongst others ( like Marathon, Chronos, Aurora, Spark etc ). When a slave becomes free it notifies the mesos master and the mesos master ( based on its [allocation](http://mesos.apache.org/documentation/latest/allocation-module/) strategy ) then creates resource offers and send them to registered frameworks. 

So in our case airflow will receive some of these resource offers from Mesos master depending upon the allocator and how many resources are free, and then its upto the Airflow's MesosExecutor whether to accept or reject the offer. In its current form, the Mesos Executor has a queue where it has list of scheduled tasks and as it receives offers from Mesos master it checks for resources in that offer and if that resources value matches the expectations in the config then it accepts the offer and launches tasks on that mesos slave via master. Hope this clears it up.

**Comment by agrajm** [comment]
@bolkedebruin please see [my comment](https://github.com/apache/incubator-airflow/pull/3008#issuecomment-368377140) above regarding how IMO we could probably add Mesos Python eggs to Travis and the time it would incur to build mesos from source for the same. Since these eggs are neither `pip` installable currently nor they are readily available for download. Do you think it would be a good idea to build mesos from source to fetch these eggs?

**Comment by sebradloff** [comment]
any update on this PR @bolkedebruin? Would really love to see this merged 😄

**Comment by bolkedebruin** [comment]
Ok merging after some small changes. See comments

File: tests/contrib/executors/__init__.py
Comment (bolkedebruin) [inline tests/contrib/executors/__init__.py]: please correct ASF header (see other files)

File: tests/contrib/executors/test_mesos_executor.py
Comment (bolkedebruin) [inline tests/contrib/executors/test_mesos_executor.py:1]: Idem

File: airflow/contrib/executors/mesos_executor.py
Comment (bolkedebruin) [inline airflow/contrib/executors/mesos_executor.py]: Can you make this cleaner? Ie. without the `\` for the multiline. Maybe use parenthesis?

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Some small nits. Given the fact that you guys are so enthusiastic about the change, maybe you want to document it in https://github.com/apache/incubator-airflow/blob/master/docs/configuration.rst#scaling-out-on-mesos-community-contributed of include an example there?

**Comment by sebradloff** [comment]
@bolkedebruin Happy to take a crack at adding the documentation. I should be able to get to that tomorrow afternoon and hope to get a rough draft up for @agrajm to review.

**Comment by sebradloff** [comment]
@agrajm I have a gist of a Mesos Dockerfile and then how I use that image and the eggs created, pulling it into my dockerized version of airflow: https://gist.github.com/sebradloff/9704e9aaa66caa8eb2873b6b390c4aac#file-dockerfile-L30

I'm happy to make the Mesos image publicly available to make it easier to install the Mesos python eggs if that helps.

**Comment by agrajm** [comment]
@sebradloff yes it would be a good to have the eggs directly available for download but since these are platform specific and mesos version dependent so it may be more useful to have the gist which explains how to fetch & install mesos eggs from mesos source. 

Also while you are at fixing documentation as suggested by @bolkedebruin can you also update the header files and other trivial changes that are asked for? Or do you want me to take a shot at it?

**Comment by sebradloff** [comment]
@agrajm Absolutely willing to fix those comments as well. How would you like me to contribute to your PR? For now I can clone your fork and checkout the branch make changes there and happy to port it to wherever you need me to 😄

**Comment by agrajm** [comment]
@sebradloff yes you can work with my fork and make changes in the AIRFLOW-2068 branch. Make sure you squash the commit with your changes and basically follow all the other [PR guidelines](https://github.com/apache/incubator-airflow/blob/master/CONTRIBUTING.md#pull-requests-guidelines) like checking for linting etc. Much thanks for doing this on such a quick notice. Do let me know if you need any help whatsoever.

**Comment by sebradloff** [comment]
@agrajm I've got the ASF headers and multiline fixes, and a stab at the documentation merged. I rebased incubator-airflow master branch on top of this and squashed. Seems that travis-ci failed on an apt-get install of some packages, but I don't see a proper error message. Let me know if this is up to your standards on the documentation or if I missed anything 😄

**Comment by sebradloff** [comment]
Looks like some packages dissapeared from apt which is why it fails:
```Package mysql-client-5.6 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  mysql-testsuite-5.7 mariadb-server-core-10.0 mariadb-client-10.0 mysql-server-core-5.7
  mysql-client-5.7

Package mysql-client-core-5.6 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  mariadb-client-core-10.0 mysql-client-core-5.7

Package mysql-server-5.6 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  percona-xtradb-cluster-server-5.6 percona-server-server-5.6 mysql-testsuite-5.7
  mariadb-server-10.0 mysql-server-core-5.7 mysql-server-5.7

Package oracle-java8-installer is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'mysql-server-5.6' has no installation candidate
E: Package 'mysql-client-core-5.6' has no installation candidate
E: Package 'mysql-client-5.6' has no installation candidate
E: Package 'oracle-java8-installer' has no installation candidate```

**Comment by sebradloff** [comment]
@agrajm rebased the most recent changes and the tests have passed. Let me know if there's anything you'd like me to change about the documentation. Long term, I'm looking forward to releasing a more comprehensive guide on best practices for deploying airflow in a containerized environment. For now, I believe the documentation here provides all the information necessary for getting airflow running with the MesosExecutor in a dockerized environment.

Once you approve I'll make sure to @ `bolkedebruin` to ask for a final review and merge 😄

**Comment by agrajm** [comment]
thanks @sebradloff for taking care of this. The changes in documentation looks sufficiently good for anyone starting up with containers and mesos. Looking forward to your comprehensive guide for deploying airflow. I'm sure it will be a good one.

**Comment by sebradloff** [comment]
@bolkedebruin We're ready for your review on the changes you requested. I'm making myself available to incorporate any changes that might still be needed in order to get this approved. 😄

File: tests/contrib/executors/__init__.py
Comment (sebradloff) [reply tests/contrib/executors/__init__.py]: Corrected

File: tests/contrib/executors/test_mesos_executor.py
Comment (sebradloff) [reply tests/contrib/executors/test_mesos_executor.py:1]: Corrected

**Comment by bolkedebruin** [comment]
@sebradloff please rebase and I'll merge it. Please be quick ;-). As part of the rebase please provide a proper commit message and description. Prepend your message according to the guidelines.

**Comment by sebradloff** [comment]
@bolkedebruin I've rebased and the updated the commit to be more inline with the commit guidelines. Please let me know if there's anything else 😃

**Pull request closed without merging**
